corehq/apps/case_importer/tasks.py

```diff
@@ -11,38 +11,35 @@
 from .exceptions import ImporterError
 from .tracking.analytics import get_case_upload_files_total_bytes
 from .tracking.case_upload_tracker import CaseUpload
+from .tracking.task_status import make_task_status_success
 from .util import get_importer_error_message, exit_celery_with_error_message
 
 
 @task(serializer='pickle', queue='case_import_queue')
 def bulk_import_async(config, domain, excel_id):
     case_upload = CaseUpload.get(excel_id)
+    result_stored = False
     try:
         case_upload.check_file()
-    except ImporterError as e:
-        return exit_celery_with_error_message(bulk_import_async, get_importer_error_message(e))
-
-    try:
         with case_upload.get_spreadsheet() as spreadsheet:
             result = do_import(spreadsheet, config, domain, task=bulk_import_async,
                                record_form_callback=case_upload.record_form)
 
         _alert_on_result(result, domain)
-
-        # return compatible with soil
-        return {
-            'messages': result
-        }
```

Comment:
Is this the "celery result"? In other words is this what celery uses to set "the result of the original task"? And you're saying that sometimes this was stored before execution of `store_task_result.delay(excel_id)`? Is that the race condition?

Is it intentional that this method previously returned this dict, and now it returns `None` or the result of `exit_celery_with_error_message(...)`?

Reply:
> And you're saying that sometimes this was stored before execution of store_task_result.delay(excel_id)? Is that the race condition?

This was eventually what was saved as the "result" of the celery task in the result backend, with a state of `SUCCESS`. The race condition I have in mind is that the task is first marked as SUCCESS and then, I think, in a second transaction the result is updated; if another process (either this `delay`'d task or a request made from the browser to refresh the case upload history) caught the result backend at exactly the right moment, it would have `state=SUCCESS` and `result=null`, and this would be permanently copied to the `CaseUploadRecord`, even if result in the celery backend is updated a split second later.

> Is it intentional that this method previously returned this dict, and now it returns None or the result of exit_celery_with_error_message(...)?

Yup, that's intentional. The reason is that previously we were using celery backend's result field as a handing-off point to be later stored in `CaseUploadRecord`. Now we only do that in the case of failure, but in the case of success we update `CaseUploadRecord` directly, and so we no longer need that to be updated in the celery result backend.